test/auth.test.ts

```diff
@@ -0,0 +1,356 @@
+        expect(workerSpy.mock.calls[1][0]).toBe(WorkerMessage.LOGIN_FAILURE);
+
+        emitterSpy.mockRestore();
+        workerSpy.mockRestore();
+      });
+
+      it('resolves on the accounts changed emitter message', async () => {
+        setTimeout(() => {
+          auth['messenger'].emit(EmitterMessage.ACCOUNTS_CHANGED, '0x1234');
+        }, 0);
+
+        await auth.login();
+      });
+
+      it('resolves on the login success worker message when token', async () => {
+        const tokenSpy = jest.spyOn(Auth.prototype, 'getToken');
+        tokenSpy.mockImplementation(() => 'asdf');
+
+        setTimeout(() => {
+          auth['messenger'].handleWorkerMessage(WorkerMessage.LOGIN_SUCCESS);
+        }, 0);
+
+        await auth.login();
+        expect(tokenSpy).toHaveBeenCalled();
+        tokenSpy.mockRestore();
+      });
+
+      it('rejects on the login success worker message when no token', async () => {
+        const tokenSpy = jest.spyOn(Auth.prototype, 'getToken');
+        tokenSpy.mockImplementation(() => null);
+        const consoleSpy = jest.spyOn(console, 'error');
+        consoleSpy.mockImplementation(() => {});
+
+        setTimeout(() => {
+          auth['messenger'].handleWorkerMessage(WorkerMessage.LOGIN_SUCCESS);
+        }, 0);
+
+        await expect(auth.login()).rejects.toThrowError();
+        expect(consoleSpy).toHaveBeenCalled();
+        consoleSpy.mockRestore();
+      });
+
+      it('rejects on the login failure worker message', async () => {
+        const consoleSpy = jest.spyOn(console, 'error');
+        consoleSpy.mockImplementation(() => {});
+
+        setTimeout(() => {
+          auth['messenger'].handleWorkerMessage(WorkerMessage.LOGIN_FAILURE);
+        }, 0);
+
+        await expect(auth.login()).rejects.toThrowError();
+        expect(consoleSpy).toHaveBeenCalled();
+        consoleSpy.mockRestore();
+      });
+    });
+  });
+
+  describe('redirect', () => {
+    let clientId = '1234';
+    let host = 'http://localhost:1738';
+    let auth: Auth;
+    let storageGetSpy: any;
+    let storageSetSpy: any;
+    let storageRemoveSpy: any;
+    let onTokenFetchedSpy: any = jest.fn();
+
+    beforeEach(() => {
+      auth = new Auth({
+        clientId,
+        host,
+        messenger: new Messenger({
+          sharedWorkerUrl: '/worker.js',
+        }),
+        _onTokenFetched: onTokenFetchedSpy,
+      });
+      storageGetSpy = jest.spyOn(Storage.prototype, 'getItem');
+      storageSetSpy = jest.spyOn(Storage.prototype, 'setItem');
+      storageRemoveSpy = jest.spyOn(Storage.prototype, 'removeItem');
+    });
+
+    afterEach(() => {
+      auth = undefined as any;
+      storageGetSpy.mockRestore();
+      storageSetSpy.mockRestore();
+      storageRemoveSpy.mockRestore();
+    });
+
+    it.each([
+      ['whatever', true],
+      [null, false],
+    ])('calls getState to determine if redirected', (state, expected) => {
```

Comment:
Minor: The curried call can also take a formatted string that also outputs the input values. It makes it easier to distinguish which case may be failing.

Reply:
great call 💡